src/types.c

```diff
@@ -11,3 +11,22 @@ bool is_record(SEXP x) {
 bool is_scalar(SEXP x) {
   return Rf_inherits(x, "vctrs_sclr");
 }
+
+enum vctrs_type vec_typeof(SEXP x) {
```

Comment:
Looks good, and this will be straight forward to extend to other base vector types (e.g. factor, Date, ...)